Adds Headcrabs!

AnturK opened the pull request:
##### Changeling Ability - Last Resort
- 20 chem cost
- Creates small silent explosion and gibs your current body.
- Spawns a headcrab - creature which will attack nearby humans or preferably infect nearby bodies with something close to alien-larva.  The changeling egg grows inside corpses into a monkey changeling after ~180 seconds.
  *Adds silent argument for explosion proc.

Original idea was it was changeling controlled but i think that would be bit too good.
Will add changelog in a moment. Also could use better sprite.

Antimattercarp commented:
What role will it fill? I find it would less than useful once a ling is on it's way to the gibber

PKPenguin321 commented:
so it can only infect corpses?

AnturK commented:
Lings usually end dead by being mobbed, If ling managed to get someone low on health this gives it another way to survive other than monkey yakety sax. Alternatively a semi-dangerous body disposal.

And yeah only corpses right now but that's easily changed.

PKPenguin321 commented on code/modules/mob/living/simple_animal/hostile/headcrab.dm:
```diff
@@ -0,0 +1,78 @@
+
+/mob/living/simple_animal/hostile/headcrab
+	name = "Headcrab"
+	desc = "Changeling last weapon"
```
this could/should be something else, maybe describing what the crab looks like

AnturK replied:
Yeah, i was out of description ideas when i wrote this

RemieRichards commented:
although it can easily be changed, it should probably remain corpses only.
(otherwise powergaming lings would literally spend 20 chems for a gib all the time)

AnturK commented:
Oh right forgot to say they can be removed by cavity implant surgery. Or you know corpse burning.

Cheridan commented:
Crainiumlobster :tm:

RemieRichards commented:
Cerebral Crustacean

GunHog commented:
AUUUUUUUUUUGGGGGGGGGHHH!!! :+1:

AlexCrimson commented:
Seems kinda pointless to limit it to corpses only. Why would you even bother using it on one? Just make the explosion have a delay and warning message like "<insert name> begins to shake violently!".

PKPenguin321 commented:
some things i noticed during testing:
+ works surprisingly well transferring the client into the monkey that's spawned
+ headcrab itself seems pretty solid
- message when headcrab infects someone is very obvious and unformatted (literally says "INFECTED Moises Pratt" in plain text), either this message shouldn't exist so it's stealthier or should be bolded in red and say something like "The creature plants something strange on Moises Pratt!" (personally i think the message shouldn't show up at all)
- removal of the changeling egg through cavity implant doesn't seem to work, tested with pretty much every other surgery and nothing gets the egg out. same with burning, i burned the infected corpse until it husked with no dice. i'm guessing these features are intended but not yet in the PR?
+/- explosion caused by gibbing yourself to make the headcrab it pretty silly, it's able to blow down nearby walls and doing about 24 brute to adjacent unarmored targets.
- said explosion doesn't have much of a chance to actually stun nearby people, meaning the headcrab doesn't get the head start it was advertised to have
- description for the power is missing a space ("... nearby.If ...")
+/- the time it takes for the egg to hatch is pretty seriously long, not sure how it plays out in a balance sense
+/- seemingly no way to detect if a corpse is infected unless you saw it get infected, personally i see this as a plus for the added stealth factor
+/- the creature itself is rather weak, dealing 5 damage per hit, hitting slowly and dying in only a few punches. if there is a corpse around to infect it doesn't matter much since it dies once it infects something anyways, but personally i think it should either do more damage or have slightly more health (not both). that or make the stun from the explosion extra long

all in all, this is almost good enough to merge as-is but could use a bit of polishing
:+1: :+1: :+1:

Nienhaus commented:
The sprite needs a lot of work, it also doesn't help it's only one direction.

Fox-McCloud commented:
why not just have it be a passive thing?

If you have the ability and enough chems/dna it just activates on death (no gibbing of the original body)--which spawns the NPC mob that attempts to latch onto the nearest target; if it's a human it infects them--if it's not removed by a set time the headcrab "takes over" that host (ie: kicks out their ckey and puts the changeling's ckey in them and gives them power/etc).

Jordie0608 commented:
Is this ready for merge?

AnturK commented:
Codewise yeah, i guess i'll rework it if it turns out too situational.

AnturK committed:
- Headcrabs

- Changelog & Name

- Description
- Removes debug message
- Fixes cavity surgery removal

Oops

- Headslug sprites courtesy of ZyloWalsh
- Squash

Giacom commented on code/game/gamemodes/changeling/powers/headcrab.dm:
```diff
@@ -0,0 +1,15 @@
+/obj/effect/proc_holder/changeling/headcrab/sting_action(var/mob/user)
+	explosion(get_turf(user),0,0,2,0,silent=1)
+	spawn(5) // So it's not killed in explosion
+		var/mob/living/simple_animal/hostile/headcrab/crab = new(get_turf(user))
```
There is a likelihood that the **user** is going to be destroyed (gibbed) after the spawn(5) starts executing. You should instead store the turf before the spawn and then use that to avoid a NullPointerException.

Giacom commented on code/modules/mob/living/simple_animal/hostile/headcrab.dm:
```diff
@@ -0,0 +1,79 @@
+			return
+	target.attack_animal(src)
+
+/obj/item/changeling_egg
```
This is duplicated from the larva code. Maybe these two classes can share an abstract class to remove the duplicated code.

AnturK replied:
Yeah that was my intention at the start, forgot i left it like this. Will make this bit neater in a moment.

AnturK committed:
- Makes body_egg parent for alien embryos/changeling eggs

AnturK commented:
Wasn't entirely sure where to put the common ancestor so i just dumped it in items dir
